File: lib/models/game.dart
import 'dart:convert';

import 'package:digital_game_shop/services/gamesApiService.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

class Game {
  String id;
  String logo;
  String title;
  String description;
  String distributor;
  String date;
  double price;
  double personalScore;
  double globalScore;
  double progress;
  List userId;

  Game(
      {this.id,
      @required this.logo = 'URL',
      @required this.title,
      this.description,
      @required this.distributor,
      @required this.date,
      @required this.price,
      this.personalScore,
      @required this.globalScore = 0,
      this.progress,
      this.userId});

  static Game fromMap(Map<String, dynamic> game) => Game(
      id: game['id'],
      logo: game['logo'],
      title: game['title'],
      description: game['description'],
      distributor: game['distributor'],
      date: game['date'],
      price: game['price'],
      personalScore: game['personal_score'],
      globalScore: game['global_score'],
      progress: game['progress'],
      userId: game['user_id']);

  Map<String, dynamic> toMap() => {
        "id": id,
        "logo": logo,
        "title": title,
        "description": description,
        "distributor": distributor,
        "date": date,
        "price": price,
        "personal_score": personalScore,
        "global_score": globalScore,
        "progress": progress,
        "user_id": userId
      };

  static List<Game> myGamesFromJson(String jsonData, String filterId) {
    final data = json.decode(jsonData);
    final games = List<Game>.from(data.map((game) => Game.fromMap(game)));
    var filteredGames = <Game>[];
    games.forEach((game) {
      if (game.userId.contains(filterId)) filteredGames.add(game);
    });
    return filteredGames;
  }

  static List<Game> gamesFromJson(String jsonData) {
    final data = json.decode(jsonData);
    final games = List<Game>.from(data.map((game) => Game.fromMap(game)));
    return games;
  }

  static fromJson(String jsonData) {
    final data = json.decode(jsonData);
    return fromMap(data);
  }

  toJson() => json.encode(toMap());

  void copyFrom(Game otherGame) {
    id = otherGame.id;
    logo = otherGame.logo;
    title = otherGame.title;
    description = otherGame.description;
    distributor = otherGame.distributor;
    date = otherGame.date;
    price = otherGame.price;
    personalScore = otherGame.personalScore;
    globalScore = otherGame.globalScore;
    progress = otherGame.progress;
    userId = otherGame.userId;
  }
}

File: backend/src/models/game.js
import mongoose, { mongo } from 'mongoose'
import validator from 'validator'

const Schema = mongoose.Schema

const GameSchema = new Schema({
    logo: {
        type: String,
        required: true,
        unique: true
    },
    title: {
        type: String,
        required: true,
        unique: false,
        trim: true
    },
    description: {
        type: String,
        required: false,
        unique: false,
        trim: true
    },
    distributor: {
        type: String,
        required: true,
        unique: false
    },
    date: {
        type: String,
        required: true,
        unique: false,
        trim: true/*,
        validate: value => {
            if (!validator.isDate(value)) {
                throw new Error({error: 'Invalid Date'})
            }
        }*/
    },
    price: {
        type: Number,
        required: true,
        unique: false
    },
    personal_score: {
        type: Number,
        required: false,
        unique: false
    },
    global_score: {
        type: Number,
        required: false,
        unique: false
    },
    progress: {
        type: Number,
        required: false,
        unique: false
    },
    user_id: {
        type: Array,
        required: false,
        unique: false
    }
}, { versionKey: false })

const Game = mongoose.model('Game', GameSchema)

export default Game